File: client/src/main/scala/stasis/client/ops/search/DefaultSearch.scala
package stasis.client.ops.search

import java.time.Instant

import scala.concurrent.ExecutionContext
import scala.concurrent.Future

import stasis.client.api.clients.ServerApiEndpointClient

class DefaultSearch(
  api: ServerApiEndpointClient
)(implicit ec: ExecutionContext)
    extends Search {
  override def search(query: Search.Query, until: Option[Instant]): Future[Search.Result] = {
    val metadataResult = for {
      definitions <- api.datasetDefinitions()
      entries <-
        Future
          .sequence(
            definitions.map { definition =>
              api.latestEntry(definition.id, until).map(entry => (definition, entry))
            }
          )
      metadata <-
        Future
          .sequence(
            entries.map {
              case (definition, Some(entry)) =>
                api.datasetMetadata(entry).map(metadata => (definition, Some((entry, metadata))))

              case (definition, None) =>
                Future.successful((definition, None))
            }
          )
    } yield {
      metadata
    }

    metadataResult.map { metadata =>
      val definitions = metadata.map {
        case (definition, Some((entry, metadata))) =>
          val matches = metadata.filesystem.entities.filter { case (path, _) =>
            query.pattern.matcher(path.toAbsolutePath.toString).matches
          }

          val result = Search.DatasetDefinitionResult(
            definitionInfo = definition.info,
            entryId = entry.id,
            entryCreated = entry.created,
            matches = matches
          )

          if (matches.nonEmpty) {
            (definition.id, Some(result))
          } else {
            (definition.id, None)
          }

        case (definition, None) =>
          (definition.id, None)
      }

      Search.Result(definitions = definitions.toMap)
    }
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/search/DefaultSearch.kt
package stasis.client_android.lib.ops.search

import stasis.client_android.lib.api.clients.ServerApiEndpointClient
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.flatMap
import stasis.client_android.lib.utils.Try.Companion.map
import stasis.client_android.lib.utils.Try.Success
import java.time.Instant

class DefaultSearch(
    private val api: ServerApiEndpointClient
) : Search {
    override suspend fun search(query: Regex, until: Instant?): Try<Search.Result> {
        val regex = query.toPattern()

        val definitions = (api.datasetDefinitions().getOrElse { emptyList() })
            .map { definition -> definition to api.latestEntry(definition.id, until) }
            .map { (definition, entry) ->
                val entryMetadata = entry.flatMap {
                    it?.let {
                        api.datasetMetadata(it).map { metadata -> it to metadata }
                    } ?: Success(null)
                }

                definition to entryMetadata
            }
            .map { (definition, tryEntryMetadata) ->
                tryEntryMetadata.map { entryMetadata ->
                    when (entryMetadata) {
                        null -> definition.id to null
                        else -> {
                            val (entry, metadata) = entryMetadata

                            val matches = metadata.filesystem.entities.filter { (path, _) ->
                                regex.matcher(path).matches()
                            }

                            val result = Search.DatasetDefinitionResult(
                                definitionInfo = definition.info,
                                entryId = entry.id,
                                entryCreated = entry.created,
                                matches = matches
                            )

                            if (matches.isNotEmpty()) {
                                definition.id to result
                            } else {
                                definition.id to null
                            }
                        }
                    }
                }
            }

        return Try(seq = definitions).map { Search.Result(definitions = it.toMap()) }
    }
}
